### PR TITLE
Add several multi-subject inverse models via Mutar

### PR DESCRIPTION
**Pull request by hichamjanati**
This PR generalizes the current GroupLasso solver to include all multi-task models of Mutar https://hichamjanati.github.io/mutar/.

### CODE REVIEW

**Comment by agramfort** [comment]
shall I review @hichamjanati ?

**Comment by hichamjanati** [comment]
Yes please :)

File: groupmne/inverse.py [hunk 1]
```diff
@@ -7,24 +7,168 @@
...
+        betamax = abs(xty).max() / n_channels
+        if method in ["lasso", "relasso"]:
+            alpha_ = betamax * np.ones(n_subjects)
+            estimator_kwargs["alpha"] *= alpha_
+        else:
+            estimator_kwargs["beta"] *= betamax
+    return estimator_kwargs
+
+
+def compute_time_dependent_gl(gains, meeg, group_info, depth=0.9,
```
Comment (agramfort) [inline groupmne/inverse.py]: is this meant to be a public function? Currently according to the doc the only public functions are:

   compute_fwd
   compute_inv_data
   compute_group_inverse

File: groupmne/inverse.py [hunk 2]
```diff
@@ -35,17 +179,12 @@ def compute_group_inverse(gains, M, group_info, method="grouplasso",
...
     return_stc : bool, (optional, default True)
         If true, source estimates are returned as stc objects, array otherwise.
     time_independent : bool, (optional, default false)
         If True, each time point is solved independently. By default,
         the group lasso is applied on the time and subjects axes.
-    n_jobs : int (default 1)
-        The number of CPUs to use in parallel.
-    kwargs : dict
+    estimator_kwargs : dict
         additional arguments passed to the solver.
```
Comment (agramfort) [inline groupmne/inverse.py]:
```
    ** estimator_kwargs
        Arbitrary keyword arguments to be passed to the solver.
```

also as you call it solver maybe estimator_kwargs -> solver_kwargs?

**Review by agramfort — COMMENTED** [review]
for MNE users this API will still be too low level. compute_group_inverse should take as input a list of forward objects and a list of evoked object.

ideally the example would then be much shorted and simpler.

load evoked for all my subjects
get a forward for all my subjects by morphing a src from fsaverage
call compute_group_inverse and get an stc for each subject.

that's it.

File: groupmne/inverse.py (hunk 1 above)
Comment (hichamjanati) [reply groupmne/inverse.py]: Yes it should be made private

**Comment by hichamjanati** [comment]
> for MNE users this API will still be too low level. compute_group_inverse should take as input a list of forward objects and a list of evoked object.

The `compute_inv_data` step that returns np arrays can be be private indeed so that the user only works with MNE objects. 
- fwds <- compute_fwds
- stcs <- compute_group_inverse(fwds, evokeds, ...)

**Comment by hichamjanati** [comment]
To hide the transition to numpy arrays and also avoid recomputing the aligned and filtered leadfields + permutations I created an InverseOperator class. Wdyt about this api @agramfort ?

File: groupmne/inverse.py [hunk 3]
```diff
@@ -7,93 +7,356 @@
...
+    fwds: list of `mne.Forward`.
+        Forward soluton of each subject.
+    noise_covs: list of `mne.Covariance`
+        Noise covariance of each subject.
+    src_ref: instance of `mne.SourceSpaces`.
+        Source space to be used as a reference for the group study. `src_ref`
+        will be morphed to the individual surface of each subject to have
+        aligned sources.
+    ch_type: str.
+        Type of channels to use. Must be one of "grad", "mag" or "eeg".
```
Comment (agramfort) [inline groupmne/inverse.py]: there is no reason to restrict things to single channel types if you pass a noise cov. I would just take any available data channel in the evokeds passed.

File: groupmne/inverse.py [hunk 4]
```diff
@@ -7,93 +7,356 @@
...
-    kwargs : dict
-        additional arguments passed to the solver.
-
-    Returns
-    -------
-    estimates: instance of SourceEstimates | array
-        The estimated sources as an array if `return_stc` is False.
-    log : dict
-        Some info about the convergence.
+    fwds: list of `mne.Forward`.
```
Comment (agramfort) [inline groupmne/inverse.py]:
```suggestion
    fwds: list of `mne.Forward`
```
remove all the dots for parameter types. Just needed to short description below.

File: groupmne/utils.py [hunk 5]
```diff
@@ -200,3 +203,67 @@ def _compute_coreg_dist(subject, trans_fname, info_fname, subjects_dir):
...
+                    # between i and j by interspacing
+                    # k somewhere along the current
+                    # path
+                    if ((d_ikj < dist[i, j]) or (dist[i, j] == 0)):
+                        dist[i, j] = d_ikj
+    return dist
+
+
+def compute_ground_metric(src, group_info):
+    """Compute geodesic distance matrix on the triangulated mesh of src."""
```
Comment (agramfort) [inline groupmne/utils.py:236]: missing docstring for a public function

Comment (agramfort) [reply groupmne/utils.py:236]: just make the function private

File: groupmne/utils.py [hunk 6]
```diff
@@ -200,3 +203,67 @@ def _compute_coreg_dist(subject, trans_fname, info_fname, subjects_dir):
...
+        A = A[verts][:, verts]
+    A = A.toarray()
+    A[A == 0.] = 1e6
+    A.flat[::len(A) + 1] = 0.
+    A = floyd_warshall(A)
+    return A
+
+
+@jit(nogil=True, cache=True, nopython=True)
+def floyd_warshall(dist):
```
Comment (agramfort) [inline groupmne/utils.py]:
```suggestion
def _floyd_warshall(dist):
```
make it private

File: groupmne/utils.py [hunk 7]
```diff
@@ -200,3 +203,67 @@ def _compute_coreg_dist(subject, trans_fname, info_fname, subjects_dir):
     dist = M[np.arange(len(info_dig)), idx].mean()
 
     return dist
+
+
+def mesh_all_distances(points, tris, verts=None):
```
Comment (agramfort) [inline groupmne/utils.py]:
```suggestion
def _mesh_all_distances(points, tris, verts=None):
```

File: examples/plot_group_lasso.py [hunk 8]
```diff
@@ -145,11 +138,13 @@ def process_meg(raw_name):
 # it must be set as a positive number between 0 and 1. With alpha = 1, all
 # the sources are 0.
 
-stcs, log = compute_group_inverse(gains, M, group_info,
-                                  method="grouplasso",
-                                  depth=0.9, alpha=0.6, return_stc=True,
-                                  n_jobs=4)
+inv_op = InverseOperator(fwds, noise_covs, src_ref, ch_type="grad",
+                         depth=0.9)
+inv_op.compute_group_model()
```
Comment (agramfort) [inline examples/plot_group_lasso.py]: I would avoid an object and try to do something like:

```
fwds = prepare_forwards(fwds, src_ref)
stcs = computer_group_inverse(fwds, evokeds, noise_covs, depth=0.9, method="grouplasso")
```

I would take the tmin, tmax from the evokeds imposing that they have the same times attribute.
I would call time_independent -> spatiotemporal= True | False

if True the solver works with both space and time.

File: groupmne/__init__.py [hunk 9]
```diff
@@ -1,12 +1,11 @@
...
-from .inverse import compute_group_inverse
+from .group_model import compute_gains, get_src_reference, compute_fwd
+from .inverse import InverseOperator
 from ._version import __version__
 
 
-__all__ = ["group_model", "inverse", "compute_gains", "compute_inv_data",
-           "get_src_reference", "compute_group_inverse", "compute_fwd",
+__all__ = ["group_model", "inverse", "compute_gains",
+           "get_src_reference", "InverseOperator", "compute_fwd",
```
Comment (agramfort) [inline groupmne/__init__.py]: only put in the public API what is used in the documentation / examples. The rest is just for you and your internal cooking.

File: examples/plot_group_lasso.py (hunk 8 above)
Comment (hichamjanati) [reply examples/plot_group_lasso.py]: The problem with this is the overhead of aligning the forwards (first aligning and then checking all vertices that were thrown away by the BEM for each subject and then compute the intersection of the vertices). Recomputing an stc with a different hyperparam or solver would require doing this all over again ... 
Ok tmin / tmax; and  spatiotemporal, I like it

File: groupmne/__init__.py (hunk 9 above)
Comment (hichamjanati) [reply groupmne/__init__.py]: Yes

**Comment by agramfort** [comment]
can you work with fwds assuming the fwds have been modified by
prepare_forwards so no further check is needed?
basically prepare_fowards will make you have new fwds that are compatible
out of the box.

>

**Comment by hichamjanati** [comment]
I'm not sure whether I can modify instances of fwds by removing some vertices and reordering them. The souvenir I have from last year's adventures (https://github.com/mne-tools/mne-python/issues/5349) is that it requires hacking all mne objects (SourceSpaces, Forward.. ) to have vertices in a -not sorted- order.

**Comment by agramfort** [comment]
you will have a problem when you save the forward to disk. Otherwise you
can do anything you
want with the forward instances especially since you control the inverse
solvers.

>

**Comment by hichamjanati** [comment]
True I forgot MNE objects are just dictionaries. Ok so we can add the gains / permutations as new attributes to fwd instances. Maybe we can avoid overwriting stuff in fwds so as not to have problems if the user calls single subject mne solvers on a modified fwd ?

**Comment by agramfort** [comment]
sure that would be a good idea.

File: docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+# Source and forward modeling
+# ---------------------------
+# To guarantee an alignment across subjects, we start by
+# computing (or reading if available) the source space of the average
+# subject of freesurfer `fsaverage`
+# If fsaverage is not available, it will be fetched to the data_path
+
+resolution = 4
+spacing = "ico%d" % resolution
+src_ref = get_src_reference(spacing=spacing, subjects_dir=subjects_dir)
```
Comment (agramfort) [inline docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]: I would not introduce this function. Just use directly MNE public functions. I would delete it.

File: docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+###################################################################
```
Comment (agramfort) [inline docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]: I would use a subsection:

```
###########################################
# Compute forward models with a reference source space
# ---------------------------------------------------------
```

File: docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+parallel, run_func, _ = parallel_func(compute_fwd, n_jobs=n_jobs)
+
+fwds = parallel(run_func(s, src_ref, info, trans, bem,  mindist=3)
+                for s, info, trans, bem in zip(subjects, raw_name_s,
+                                               trans_fname_s, bem_fname_s))
+
+fwds = prepare_fwds(fwds, src_ref)
+evokeds = [ev.crop(0.015, 0.025).copy()
+           for ev in evokeds]
+# We can now compute the data of the inverse problem.
```
Comment (agramfort) [inline docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]: remove this line

File: docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+# We restric the time points around 20ms in order to reconstruct the sources of
```
Comment (agramfort) [inline docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]: say this were you do it

File: docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+# the N20 response.
+
+###########################################
+# Solve the inverse problems with groupmne
```
Comment (agramfort) [inline docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]:
```suggestion
# Solve the inverse problems with group-Lasso
```

actually I would prefer to call this multi-task Lasso

File: docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+
+stcs = compute_group_inverse(fwds, evokeds, noise_covs,
+                             method='grouplasso',
+                             spatiotemporal=True,
+                             alpha=0.8)
+
+############################################
+# Let's visualize the N20 response. The stimulus was applied on the right
+# hand, thus we only show the left hemisphere. The activation is exactly in
+# the Primary somatosensory cortex. We highlight the borders of the post
```
Comment (agramfort) [inline docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]:
```suggestion
# the primary somatosensory cortex. We highlight the borders of the post
```

File: docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+t_idx = stcs[0].time_as_index(t)
+view = "lateral"
+
+for stc, subject in zip(stcs, subjects):
+    g_post_central = mne.read_labels_from_annot(subject, "aparc.a2009s",
+                                                subjects_dir=subjects_dir,
+                                                regexp="G_postcentral-lh")[0]
+    n_sources = [stc.vertices[0].size, stc.vertices[1].size]
+    m = abs(stc.data[:n_sources[0], t_idx]).max()
+    surfer_kwargs = dict(
```
Comment (agramfort) [inline docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]: calling this surfer is not future proof with new pyvista. Call this plot_kwargs ?

File: docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+from mne.inverse_sparse import mixed_norm  # noqa: E402
+
+t = 0.02
+t_idx = stcs[0].time_as_index(t)
+view = "lateral"
+for subject, evoked, fwd, cov in zip(subjects, evokeds, fwds, noise_covs):
+    ev = evoked.copy()
+    ev.pick_types(meg="grad")
+    ev.crop(0.015, 0.025)
+    stc = mixed_norm(ev, fwd, cov, alpha=80., loose=0.)
```
Comment (agramfort) [inline docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]: why not just calling compute_group_inverse with one subject? are you sure it gives the same result at mixed_norm with loose=0?

File: docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+    stc.subject = subject
+    g_post_central = mne.read_labels_from_annot(subject, "aparc.a2009s",
+                                                subjects_dir=subjects_dir,
+                                                regexp="G_postcentral-lh")[0]
+    n_sources = [stc.vertices[0].size, stc.vertices[1].size]
+    m = abs(stc.data[:n_sources[0], t_idx]).max()
+    surfer_kwargs = dict(
```
Comment (agramfort) [inline docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]: plot_kwargs?

Comment (agramfort) [reply docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]: actually you could avoid a lot of code duplication with this kwargs and highlight was kwargs change between plots

File: docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+        smoothing_steps=5, cortex=("gray", -1, 6, True))
+    brain = stc.plot(**surfer_kwargs, views=view)
+    brain.add_text(0.1, 0.9, subject + "_mxne", "title")
+    brain.add_label(g_post_central, borders=True, color="green")
+
+
+###########################################
+# References
+# ----------
+# [1] Lim et al, Sparse EEG/MEG source estimation via a group lasso, PLOS ONE,
```
Comment (agramfort) [inline docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]: put all authors as you did for [2]

File: docs/_downloads/ee38530357642bb04ac0e6b7c2b97e61/plot_group_lasso.py [hunk 20]
```diff
@@ -0,0 +1,219 @@
+"""
+Multi-subject joint source localization with multi-task models.
+===============================================================
```
Comment (agramfort) [inline docs/_downloads/ee38530357642bb04ac0e6b7c2b97e61/plot_group_lasso.py]: remove this file

File: docs/_downloads/ee38530357642bb04ac0e6b7c2b97e61/plot_group_lasso.py [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+
+###########################################
+# References
+# ----------
+# [1] Lim et al, Sparse EEG/MEG source estimation via a group lasso, PLOS ONE,
+# 2017
+#
+# [2] Jussi Nurminen, Hilla Paananen, & Jyrki Mäkelä. (2017). High frequency
+# somatosensory MEG: evoked responses, FreeSurfer reconstruction [Data set].
+# Zenodo. http://doi.org/10.5281/zenodo.889235
```
Comment (agramfort) [inline docs/_downloads/ee38530357642bb04ac0e6b7c2b97e61/plot_group_lasso.py]: all the files you generated in docs like the pngs should not be added to the repo. Just below...

File: docs/_modules/groupmne/inverse.html [hunk 22]
```diff
@@ -0,0 +1,411 @@
+<!DOCTYPE html>
```
Comment (agramfort) [inline docs/_modules/groupmne/inverse.html]: remove file

**Review by agramfort — COMMENTED** [review]
that's a start. Please cleanup all the files that are generated by sphinx

File: docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py (hunk 17 above)
Comment (hichamjanati) [reply docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]: That's a good sanity check, I'll do it

Comment (hichamjanati) [reply docs/_downloads/433d4b2ff389555700f26fe734f79b9a/plot_group_lasso.py]: Yes :) support and amplitudes,  max diff = 1e-5. We can make it a test.

**Comment by agramfort** [comment]
Please make it a test

File: groupmne/tests/test_inverse.py [hunk 23]
```diff
@@ -1,33 +1,217 @@
...
+#
+#     support_gl = np.where(stc_gl.data.any(1))[0]
+#     support_mx = np.where(stc_mx.data.any(1))[0]
+#     ampl_gl = stc_gl.data[support_gl]
+#     ampl_mx = stc_mx.data[support_mx]
+#     # vertices_gl = stc_gl.vertices[0][support_gl]
+#     # vertices_mx = stc_mx.vertices[0][support_mx]
+#
+#     # assert_array_equal(vertices_gl, vertices_mx)
+#     assert abs(ampl_gl - ampl_mx).max() < 1e-4
```
Comment (agramfort) [inline groupmne/tests/test_inverse.py:217]: Why commenting the test ?

Comment (hichamjanati) [reply groupmne/tests/test_inverse.py:217]: For some reason It works with the example HF data but not with the testing data, I get slightly different sources. The preprocessing of the cov in groupmne is based the one in  the sovlers of `mne.minimum_norm` but I noticed it is not the same in the sparse solvers`mne.inverse_sparse`. I need to do some digging offline before attempting this test

**Comment by agramfort** [comment]
ok. Shall we merge this one as is or do it in this PR?

>

**Comment by hichamjanati** [comment]
I vote for merging this one once it's green to keep PRs independent + we need to discuss the preprocessing in mne before doing anything

**Comment by agramfort** [comment]
ok green !

**Pull request merged**